File: 31_oct/problem_07_max_product_alternating/solution.py
class Solution(object):
    def maxProduct(self, nums, k, limit):
        """
        :type nums: List[int]
        :type k: int
        :type limit: int
        :rtype: int
        """
        from functools import lru_cache
        
        n = len(nums)
        
        @lru_cache(maxsize=None)
        def dp(idx, current_sum, length):
            # Base case
            if idx == n:
                if current_sum == k and length > 0:
                    return 1
                return -1
            
            # Skip current element
            result = dp(idx + 1, current_sum, length)
            
            # Take current element
            sign = 1 if length % 2 == 0 else -1
            new_sum = current_sum + sign * nums[idx]
            sub_result = dp(idx + 1, new_sum, length + 1)
            
            if sub_result != -1:
                product = sub_result * nums[idx]
                if product <= limit:
                    result = max(result, product)
            
            return result
        
        return dp(0, 0, 0)

def solve(infile, outfile):
    """Standalone execution for orchestrator"""
    n, k, limit = map(int, infile.readline().split())
    nums = list(map(int, infile.readline().split()))
    
    solution = Solution()
    result = solution.maxProduct(nums, k, limit)
    outfile.write(str(result) + '\n')

if __name__ == "__main__":
    import sys
    solve(sys.stdin, sys.stdout)

File: 31_oct/problem_07_max_product_alternating/solution.cpp
#include <iostream>
#include <vector>
#include <map>
#include <tuple>
#include <functional>
using namespace std;

class Solution {
public:
    int maxProduct(vector<int>& nums, int k, int limit) {
        int n = nums.size();
        // dp[i][sum][parity] = max product using first i elements, 
        // with alternating sum 'sum', and parity (0=even, 1=odd length)
        map<tuple<int, int, int>, int> dp;
        
        function<int(int, int, int)> solve = [&](int idx, int sum, int parity) -> int {
            if (idx == n) {
                return (sum == k && parity == 1) ? 1 : -1;
            }
            
            auto key = make_tuple(idx, sum, parity);
            if (dp.count(key)) return dp[key];
            
            // Skip current element
            int result = solve(idx + 1, sum, parity);
            
            // Take current element
            int sign = (parity == 0) ? 1 : -1;
            int newSum = sum + sign * nums[idx];
            int subResult = solve(idx + 1, newSum, 1 - parity);
            
            if (subResult != -1) {
                long long product = (long long)subResult * nums[idx];
                if (product <= limit) {
                    result = max(result, (int)product);
                }
            }
            
            return dp[key] = result;
        };
        
        return solve(0, 0, 0);
    }
};

int main() {
    int n, k, limit;
    cin >> n >> k >> limit;
    
    vector<int> nums(n);
    for (int i = 0; i < n; i++) {
        cin >> nums[i];
    }
    
    Solution solution;
    int result = solution.maxProduct(nums, k, limit);
    cout << result << endl;
    
    return 0;
}
